managers/opportunity/index.js

```diff
@@ -48,16 +59,47 @@ class opportunityManager {
       }
       console.log('Values in QueryString', queryObject);
 
-      let fetchedOpportunitiesQuery = this.opportunity.find(queryObject);
-      fetchedOpportunitiesQuery.select('-__v');
-      fetchedOpportunitiesQuery.select('-_id');
+      const default_page_limit= 20 ;
+      const default_page_offset= 0 ;
+
+       let { offset, queryLimit} = queryObject ;
+
+
+       // Conditionals are used to handle the cases if user enters negative data in offset or ueryLimit , in that
+       // the output will be the default data (offset=0 and queryLimit=20) 
+
+       // else it will work as specified above 
+
+       // This code doesn't handle the case if User enters offset more than the max_data available 
+
+
+       
 
-      return fetchedOpportunitiesQuery;
+      if(queryLimit==undefined || queryLimit<0 || offset< 0)
+      {
+        queryLimit= default_page_limit ;
+      }
+
+      if(offset==undefined || offset<0 || queryLimit<0)
+      {
+        offset= default_page_offset ;
+      }
```

Comment:
- what if I pass a huge value to `quaryLimit`? how does that will work? will it slow down DB operation or crash it completely? 
- I think these `if` conditions are not correct. should the value of `offset` affect `queryLimit`? 
- let the user know when there is a 'next page' available. for that, you have to modify the output format a bit.

Reply:
okay as per this :   
1 .   I haven't worked with huge databases but as far as queryLimit being very high , I can see that this is an O(limit) operation
as it will try and get all the data , in case of data sets above 10^8/9  it will not work in the time limit provide, but I don't think we can improve on this as we aren't performing any operation on the range (else we could have used Segment/ Fenwick trees to make each query O(Logn) compared to O(n) which is right now ), 
a Preventive measure that I can think of in real world cases will be having multiple servers / instances to handle very big amount of queries (ie 10^5 or more ) or in else in normal case it will work fine as queries are very limited  

2. Offset here denotes the starting position from where I Start fetching the data , it will affect the limit in those cases when  (offset+ limit > dataset size ) for that we can make a check on limit tot bring it down to (dataSetsize- offset), I m not sure about exact behaviour of mongoose limit but I m guessing it does that by default  and if negative or undefined values are considered , it is already handled . 

3.  okay , I will have to think around this a bit but most probably it will be something like keeping tracking of current position of the user in the database and if the user is on the last page (then return something like no more pages ahead  or page limit reached ) and in other cases just simply make a next page available option valid ,

But I have a doubt what is next page exactly here as we are dealing in terms of database index or is it some sort of other page ?

Reply:
>  I haven't worked with huge databases but as far as queryLimit being very high ... 

my concern was about the user's input. what if the user intentionally puts huge numbers to trick the system? I think we could set a maximum offset. 

> But I have a doubt what is next page exactly here .... 

for example, if the offset is 0 and limit is 10, and there are more than 30 entries in the database, we should somehow let the client know that there are more data that could be fetched on subsequent request. 

Following is a snippet from a codebase I am working on, where I have implemented pagination. Go through it and see if you could find something valuable. If you find any issue with this snippet, please don't hesitate to point it out. 

```typescript
const RESULT_PER_PAGE = 10;

interface IOptions {
  page?: number;
}

async function joinFeedAndItem(
  options?: IOptions
): Promise<{ data: any[]; page: number; nextPage: number | null }> {
  const page = getValidPageNum(options?.page);
  const response = [];
  const feedItems = await FeedItemModel.find()
    .skip(RESULT_PER_PAGE * page)
    .limit(RESULT_PER_PAGE)
    .exec();

  const nextPage: number | null =
    (await FeedItemModel.countDocuments()) >
    RESULT_PER_PAGE * page + RESULT_PER_PAGE
      ? page + 1
      : null;

  for (let i = 0; i < feedItems.length; i++) {
    const fItem = feedItems[i].toObject();
    const feed = (
      await FeedModel.findOne({
        item: fItem._id,
      })
    ).toObject();

    (fItem as any).feed = feed || 'feed missing';
    (fItem as any).feed.itemCount = (feed as any).item?.length;

    delete (feed as any).item;
    response.push(fItem);
  }

  return { data: response, page, nextPage };
}

function getValidPageNum(num: number) {
  return (!Number.isNaN(num) && Math.max(Math.min(1000, num), 0)) || 0;
}
```

Reply:
Okay , I get your point now, 
 So let's suppose the user tries to trick the database, for example let's consider a DB with size 100 , user sets offset as 50 and limit as 100 , here we have following options :

1: Return a list with all the entries with 50 to 100
2 : don't do anything , just return an error that it is an invalid request 
3 : we can return a list of all the entries from 50 to 100 along with a message that no more pages can be fetched and 
    also in our server turning his request size from 100 to 50 only (max_size- offset)  always and keeping a conditional 
    (limit <=max_size- offset ? valid : false  && (offset validated before by either making it default or returning an error )  ) 
  to  make sure it doesn't exceed ever 
4 : For handle out of bound queries by just setting it to some default , but it has a downside that the server is always busy 
   in fetching down data even for invalid queries 

As of now I can think of all above options that can be implemented depending upon various scenarios ,

Yes, your above code does help me in understanding the the implementation of getting the next page but can you tell me which option to go for of the ones I  mentioned or any other if there to handle invalid/tricky query parameters

Reply:
respond with a valid error message. that should be enough.

Reply:
okay

Reply:
Hello @abdus sir, sorry for the delay in the nex PR , as currently I am figuring out few things that should be known before implementing this properly , some new concepts discovered, so I just need a few extra days to get it done.

Reply:
Hello @abdus @Manvityagi  Please check the Updated Code of Get Opportunities Function, 

Changes Made :

1. Now the function  tells whether there is a previous  and next Page available 
2. It is able to handle Out of Bound Queries 
3. For default I have set Page=1 and Limit=1 

Sorry for the Delay, I was having issues with the concept of async wait and how to wait , so I had to go through it a bit , therefore it took time 

Please review it and provided necessary Test cases / Edits if any , I will happy to get them this time around as soon as Possible